app/src/main/kotlin/com/simplemobiletools/calculator/activities/MoneyActivity.kt

```diff
@@ -42,22 +55,134 @@ class MoneyActivity : SimpleActivity(), Calculator , TaxCalculator {
+                location.longitude, location.latitude
+        )
+        Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show()
+    }
+
+    override fun onStatusChanged(s:String, i:Int, b:Bundle) {}
+
+    override fun onProviderDisabled(s:String) {
+        Toast.makeText(applicationContext,"Location service has been turned off", Toast.LENGTH_LONG).show()
+    }
+    override fun onProviderEnabled(s:String) {
+        Toast.makeText(applicationContext,"Locattion service has been turned on", Toast.LENGTH_LONG).show()
+    }
+
+    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+        if(requestCode == GPS_REQUEST_CODE) {
+            if(grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                Log.e("PermissionTag","Permission Granted (if 0 then is granted): " + grantResults[0])
+            }
+        }
+    }
+
+    fun taxLocationStrat() {
+        var gps_enabled : Boolean = false;
+        try{
+            gps_enabled = locationManager!!.isProviderEnabled(LocationManager.GPS_PROVIDER)
+        } catch(e: Exception) {}
+
+        //if location service is enabled then retrieve the latest location and performs tax rate with that location
+        if(gps_enabled) {
+            //Checks if location service has been granted again
+            if (ActivityCompat.checkSelfPermission(this, android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                ActivityCompat.requestPermissions(this, arrayOf(android.Manifest.permission.ACCESS_FINE_LOCATION), GPS_REQUEST_CODE);
+            }
+            var lastLocation = locationManager!!.getLastKnownLocation(LocationManager.GPS_PROVIDER)
+
+            if (lastLocation != null) {
+                currentLongitude = lastLocation.longitude
+                currentLatitude = lastLocation.latitude
+                var message: String = String.format("Current Location \n Longitude: " + lastLocation.longitude + " \n Latitude: " + lastLocation.latitude)
+                Toast.makeText(applicationContext, message, Toast.LENGTH_LONG).show()
+            }
+
+            val geocoder = Geocoder(applicationContext, Locale.getDefault())
+            val addresses: List<Address> = geocoder.getFromLocation(currentLatitude, currentLatitude, 1)
+            var province : String = ""
+            try {
+                 province = addresses.get(0).adminArea
+            } catch (e: Exception) {
+                e.printStackTrace()
+                Toast.makeText(applicationContext, "Unable to recognize your location, please choose one of the following provinces", Toast.LENGTH_SHORT).show()
+                spawnTaxModal()
+                return
+            }
+
+//            //TODO: refactor this part to make it cleaner
```

Comment:
Good call on that

Reply:
Refactoring will be done in my next commit.